ompi/mca/coll/solo/coll_solo_bcast.c

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright (c) 2019      The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "coll_solo.h"
+
+int mca_coll_solo_bcast_intra(void *buff, int count,
+                              struct ompi_datatype_t *dtype, 
+                              int root,
+                              struct ompi_communicator_t *comm, 
+                              mca_coll_base_module_t * module)
+{
+    if (ompi_datatype_is_contiguous_memory_layout(dtype, count)) {
+        mca_coll_solo_bcast_linear_intra_memcpy(buff, count, dtype, root, comm, module);
+    }
+    else {
+        mca_coll_solo_bcast_linear_intra_osc(buff, count, dtype, root, comm, module);
+    }
+    return OMPI_SUCCESS;
+}
+
+int mca_coll_solo_bcast_linear_intra_memcpy(void *buff, int count,
+                                            struct ompi_datatype_t *dtype, 
+                                            int root,
+                                            struct ompi_communicator_t *comm, 
+                                            mca_coll_base_module_t * module)
+{
+    mca_coll_solo_module_t *solo_module = (mca_coll_solo_module_t *) module;
+
+    int rank = ompi_comm_rank(comm);
+    int id;
+    ptrdiff_t extent, lower_bound;
+    ompi_datatype_get_extent(dtype, &lower_bound, &extent);
+    /* Enable solo module if necessary */
+    if (!solo_module->enabled) {
+        mca_coll_solo_lazy_enable(module, comm);
+    }
+    char *data_buf;
+    if ((size_t) count * extent <= mca_coll_solo_component.static_block_size) {
+        data_buf = solo_module->data_bufs[root];
+    } else if ((size_t) count * extent <= mca_coll_solo_component.mpool_large_block_size) {
+        if (rank == root) {
+            id = mca_coll_solo_mpool_request(mca_coll_solo_component.solo_mpool, count * extent);
+        }
+        mca_coll_solo_bcast_linear_intra_memcpy(&id, 1, MPI_INT, root, comm, module);
+        data_buf = mca_coll_solo_mpool_calculate(mca_coll_solo_component.solo_mpool, id, 
+                                                 count * extent);
+    } else {
+        /* TODO: Add support for very large messages */
+        //printf("TOO BIG\n");
```

Comment:
Ok, I get it. how can we handle this cleanly?

Reply:
Yeah, I will fix the support this week. Thanks for the review.